src/java.base/share/classes/jdk/internal/foreign/StringSupport.java

```diff
@@ -202,21 +200,77 @@ private static int strlen_byte(MemorySegment segment, long start) {
+     *
+     * @param segment to examine
+     * @param start   from where examination shall begin
+     * @throws IllegalArgumentException if the examined region contains no zero shorts
+     *                                  within a length that can be accepted by a String
+     */
+    public static int chunked_strlen_short(MemorySegment segment, long start) {
+
+        // Handle the first unaligned "head" bytes separately
+        int headCount = (int)SharedUtils.remainsToAlignment(segment.address() + start, Long.BYTES);
+
+        int offset = 0;
+        for (; offset < headCount; offset += Short.BYTES) {
             short curr = segment.get(JAVA_SHORT, start + offset);
             if (curr == 0) {
                 return offset;
             }
         }
+
+        // We are now on a long-aligned boundary so this is the "body"
+        int bodyCount = bodyCount(segment.byteSize() - start - headCount);
+
+        for (; offset < bodyCount; offset += Long.BYTES) {
+            // We know we are `long` aligned so, we can save on alignment checking here
+            long curr = segment.get(JAVA_LONG_UNALIGNED, start + offset);
+            // Is this a candidate?
+            if (mightContainZeroShort(curr)) {
+                for (int j = 0; j < Long.BYTES; j += Short.BYTES) {
+                    if (segment.get(JAVA_SHORT_UNALIGNED, start + offset + j) == 0) {
+                        return offset + j;
+                    }
+                }
+            }
+        }
+
+        // Handle the "tail"
+        return requireWithinArraySize((long) offset + strlen_short(segment, start + offset));
+    }
+
+    private static int strlen_short(MemorySegment segment, long start) {
+        // Do an initial read using aligned semantics.
+        // If this succeeds, we know that all other subsequent reads will be aligned
+        if (segment.get(JAVA_SHORT, start) == (short)0) {
+            return 0;
+        }
+        for (int offset = Short.BYTES; offset < ArraysSupport.SOFT_MAX_ARRAY_LENGTH; offset += Short.BYTES) {
+            short curr = segment.get(JAVA_SHORT_UNALIGNED, start + offset);
+            if (curr == (short)0) {
+                return offset;
+            }
+        }
         throw newIaeStringTooLarge();
     }
 
-    private static int strlen_int(MemorySegment segment, long start) {
-        // iterate until overflow (String can only hold a byte[], whose length can be expressed as an int)
-        for (int offset = 0; offset >= 0; offset += 4) {
-            int curr = segment.get(JAVA_INT, start + offset);
+    // The gain of using `long` wide operations for `int` is lower than for the two other `byte` and `short` variants
+    public static int strlen_int(MemorySegment segment, long start) {
+        // Do an initial read using aligned semantics.
+        // If this succeeds, we know that all other subsequent reads will be aligned
+        if (segment.get(JAVA_INT, start) == 0) {
```

Comment:
I would expect a simple loop here? What's the deal with avoiding alignment? If that's to speed up by avoiding alignment checks, it should not be required, as the segment being accessed is always at the same base address and the alignment check should be hoisted.

Reply:
The proposed solution retains the same semantics as the old method where we would get an alignment error if presented with an unaligned MS (e.g. `alignedSegment.asSlice(1)`) but with improved performance.

Reply:
Ok, I understand you might see improved numbers. What I'm saying is that in principle this trick should not be necessary: the loop is a counted loop, and we know we're incrementing by 8 on each iteration. So the alignment check should be hoist-able as once we check the base address there's nothing more to check.

Reply:
So, I'd suggest to capture this in a small benchmark, and perhaps file a C2 compiler issue.

Reply:
I wonder if optimisation issues can be related to using int counted loops instead of long, and using `ArraysSupport.SOFT_MAX_ARRAY_LENGTH`, maybe limit should be memory segment size. Adding large int to start can theoretically overflow long.